graph/src/blockchain/firehose_block_stream.rs

```diff
@@ -0,0 +1,252 @@
+
+                        Poll::Pending => {
+                            trace!(
+                                self.ctx.logger,
+                                "Connection is still pending when being wake up"
+                            );
+
+                            return Poll::Pending;
+                        }
+                    }
+                }
+
+                BlockStreamState::Connected(streaming) => match streaming.poll_next_unpin(cx) {
+                    Poll::Ready(Some(Ok(b))) => {
+                        match self.ctx.mapper.to_block_stream_event(&b, &self.ctx.filter) {
+                            Ok(event) => {
+                                return Poll::Ready(Some(Ok(event)));
+                            }
+                            Err(e) => {
+                                error!(
+                                    self.ctx.logger,
+                                    "Mapping block to BlockStreamEvent failed {}", e
+                                );
+                                self.state = BlockStreamState::Disconneted;
+
+                                return self.schedule_error_retry(cx);
+                            }
+                        }
+                    }
+
+                    Poll::Ready(Some(Err(e))) => {
+                        error!(self.ctx.logger, "Stream disconnected from endpoint {}", e);
+                        self.state = BlockStreamState::Disconneted;
+
+                        return self.schedule_error_retry(cx);
+                    }
+
+                    Poll::Ready(None) => {
+                        error!(self.ctx.logger, "Stream has terminated blocks range, we expect never ending stream right now");
+                        self.state = BlockStreamState::Disconneted;
+
+                        return self.schedule_error_retry(cx);
+                    }
+
+                    Poll::Pending => {
+                        trace!(
+                            self.ctx.logger,
+                            "Stream is pending, no item available yet will being wake up"
+                        );
+
+                        return Poll::Pending;
+                    }
+                },
+            }
+        }
+    }
+}
+
+impl<C: Blockchain> FirehoseBlockStream<C> {
+    /// Schedule a delayed function that will wake us later in time. This implementation
+    /// uses an exponential backoff strategy to retry with incremental longer delays.
+    fn schedule_error_retry<T>(&mut self, cx: &mut Context<'_>) -> Poll<Option<T>> {
+        self.connection_attempts += 1;
+        let wait_duration =
+            std::time::Duration::from_secs(wait_duration_secs(self.connection_attempts.clone()));
+
+        let waker = cx.waker().clone();
+        tokio::spawn(async move {
+            tokio::time::sleep(wait_duration).await;
+            waker.wake();
+        });
+
+        Poll::Pending
+    }
+}
+
+fn wait_duration_secs(attempt_number: u64) -> u64 {
```

Comment:
It's better to return a `Duration` here, then there's no confusion what unit is used for the return value